PRP: K8s Image Extractor

lanced00m opened the pull request:
This is regarding the PRP https://github.com/google/osv-scalibr/issues/995

erikvarga commented on extractor/filesystem/containers/k8simage/k8simage.go:
```diff
@@ -0,0 +1,261 @@
+			Locations: []string{input.Path},
+			Name:      name,
+			Version:   version,
+			PURLType:  purl.TypeDocker,
```
This is not really docker - can you add a new PURL type to purl.go?

mariussteffens commented on extractor/filesystem/containers/k8simage/testdata/not-yaml.yaml:
```diff
@@ -0,0 +1,22 @@
+apiVersion: apps/v1
```
Is there a reason to make this look very close like a yaml file? 
Would it be more obvious if this just does not follow the structure at all?

Right now this looks like a reasonable file that probably has some parsing error somewhere. But it is not immediately obvious (as in when you glance over the file) where that error comes from.

mariussteffens commented on extractor/filesystem/containers/k8simage/k8simage_test.go:
```diff
@@ -0,0 +1,286 @@
+		cfg          k8simage.Config
+		wantPackages []*extractor.Package
+	}{
+		{
```
how about a testcase that checks for the default behavior if not tag is set in the image property? As per your code and the spec this should return "latest" as tag.

mariussteffens commented on extractor/filesystem/containers/k8simage/k8simage_test.go:
```diff
@@ -0,0 +1,286 @@
+				},
+			},
+		},
+		{
```
How about a testcase with `containers` containing malformed or missing sub elements? Or would this lead to parsing error?

mariussteffens commented on extractor/filesystem/containers/k8simage/testdata/StatefulSet.yaml:
```diff
@@ -0,0 +1,100 @@
+apiVersion: apps/v1
```
Can we reduce most of the testfiles to be minimal? I.e., only contain those values and properties that are absolutely necessary? 

This greatly improves test readability and maintainability. We can still have on test with bloated properties to make sure that the parser does not error out on real-world files, however, for most testcases we should favor minimal testcases to reduce reviewing and readability bloat.

lanced00m commented:
@mariussteffens, thank you for reviewing this PR!
Since my PR had several mistakes, I rebuilt the extractor, tests, and test files while addressing all comments together. Please let me know if you notice any remaining issues in the next review, especially from the first round, in case I missed something.

erikvarga commented on extractor/filesystem/containers/k8simage/k8simage.go:
```diff
@@ -0,0 +1,273 @@
+)
+
+// K8sResource represents a Kubernetes resource with the fields needed for image extraction.
+type K8sResource struct {
```
Since these structs are only used in this package they can be private (i.e. start with lowercase)
Same for the ones below.

erikvarga commented on extractor/filesystem/containers/k8simage/k8simage.go:
```diff
@@ -0,0 +1,273 @@
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	// Only consider YAML/YML files
+	path := api.Path()
+	ext := strings.ToLower(filepath.Ext(filepath.Base(path)))
```
no need for filepath.Base if you also call filepath.Ext

erikvarga commented on extractor/filesystem/containers/k8simage/k8simage.go:
```diff
@@ -0,0 +1,273 @@
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) (inventory.Inventory, error) {
+	// Defer context cancellation check
+	defer func() {
+		if err := ctx.Err(); err != nil {
```
Should be fine to remove this - there's a separate ctx handling where Extract() is called so we only need context cancellation inside busy loops in Extract()

erikvarga commented on purl/purl.go:
```diff
@@ -53,6 +53,8 @@ const (
 	TypeDebian = "deb"
 	// TypeDocker is a pkg:docker purl.
 	TypeDocker = "docker"
+	// TypeK8sDocker is a pkg:docker purl.
```
Why docker?
Rename to just TypeK8s and the comment to "// TypeK8s is a pkg::k8s purl"
